### PR TITLE
feat(opencode): add OC-001 to OC-003 opencode.json validation rules

### LINKED ISSUE #300: New OpenCode support: OC-001 to OC-003 (opencode.json validation)
## Summary

Add initial validation support for OpenCode (S-tier tool). Currently has 0 dedicated rules despite being S-tier.

### OC-001: Invalid share mode (HIGH)
In `opencode.json`, `share` must be `"manual"`, `"auto"`, or `"disabled"`.

### OC-002: Invalid instruction path (HIGH)
Paths in `instructions` array must exist on disk or be valid URLs/globs. Broken instruction paths silently produce an incomplete context.

### OC-003: opencode.json parse error (HIGH)
`opencode.json` must be valid JSON/JSONC. Report parse errors with line/column location.

## Scope

- Add `FileType::OpenCodeConfig` variant to detect `opencode.json`
- Add `OpenCodeValidator` implementing the `Validator` trait
- Register in `ValidatorRegistry::register_defaults()`
- Add rules to `knowledge-base/rules.json` and `VALIDATION-RULES.md`
- Add test fixtures

## References

- [OpenCode Configuration](https://opencode.ai/docs/config)
- `agent-knowledge/agent-config-optional-fields.md`

### PR DESCRIPTION
**Pull request by avifenesh**
## Summary

Closes #300

Adds initial OpenCode support (S-tier tool) with three HIGH-severity validation rules for `opencode.json`:

- **OC-001**: Invalid share mode - `share` field must be `"manual"`, `"auto"`, or `"disabled"`
- **OC-002**: Invalid instruction path - paths in `instructions` array must exist on disk or be valid globs
- **OC-003**: opencode.json parse error - file must be valid JSON/JSONC with line/column error reporting

### Changes

**Core implementation:**
- `crates/agnix-core/src/rules/opencode.rs` - `OpenCodeValidator` implementing `Validator` trait
- `crates/agnix-core/src/schemas/opencode.rs` - Schema helpers with JSONC comment stripping, parsing, validation
- `crates/agnix-core/src/lib.rs` - `FileType::OpenCodeConfig` variant, detection, registry registration
- `crates/agnix-core/src/config.rs` - `opencode` category toggle, `OC-` prefix mapping

**Rules data:**
- `knowledge-base/rules.json` - Added 3 OC rules with full evidence metadata (total: 124 -> 127)
- `crates/agnix-rules/rules.json` - Synced copy
- `knowledge-base/VALIDATION-RULES.md` - Added OPENCODE RULES section

**Tests and fixtures:**
- 25+ unit tests in `opencode.rs` covering all rules, edge cases, config integration
- 14 unit tests in `schemas/opencode.rs` for JSONC parsing and validation
- `tests/fixtures/opencode/` - Valid fixture
- `tests/fixtures/opencode-invalid/` - Invalid fixture
- Updated parity test counts and categories in `rule_parity.rs`, `sarif.rs`, `mcp_tests.rs`

**i18n:**
- `locales/en.yml`, `es.yml`, `zh-CN.yml` - Messages and suggestions for all 3 rules

**Website docs:**
- Generated `oc-001.md`, `oc-002.md`, `oc-003.md` rule pages

## Test plan

- [x] `cargo test` passes (all test suites, 0 failures)
- [x] Rule parity tests pass (rules.json <-> VALIDATION-RULES.md <-> source code <-> SARIF <-> fixtures <-> website)
- [x] OC-001 validates share mode enum values
- [x] OC-002 validates instruction paths (glob syntax and literal path existence)
- [x] OC-003 validates JSON/JSONC parsing with error location reporting
- [x] JSONC comment stripping handles single-line, multi-line, and strings with embedded slashes
- [x] Rules respect `LintConfig` disabled_rules and category toggles

### CODE REVIEW

File: locales/en.yml [hunk 1]
```diff
@@ -396,6 +396,17 @@ rules:
     suggestion: "Create .cursor/rules/ directory and add .mdc files with frontmatter for better organization"
     legacy_empty: "Cursor rules file is empty"
 
+  # --- OpenCode (opencode.rs) ---
+  oc_001:
+    message: "Invalid share mode '%{value}'. Valid values: 'manual', 'auto', 'disabled'"
+    suggestion: "Use 'manual', 'auto', or 'disabled' for the share field"
+  oc_002:
+    message: "Instruction path '%{path}' does not exist and is not a valid glob pattern"
+    suggestion: "Fix or remove the broken instruction path"
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply locales/en.yml]: Fixed. OC-002 now uses specific message keys for each failure mode: `not_found` for missing paths, `invalid_glob` for bad glob syntax, `traversal` for absolute/`..` paths, and `type_error` for wrong field types. All three locale files updated.

File: locales/es.yml [hunk 2]
```diff
@@ -327,6 +327,17 @@ rules:
...
+  oc_001:
+    message: "Modo de compartir invalido '%{value}'. Valores validos: 'manual', 'auto', 'disabled'"
+    suggestion: "Usa 'manual', 'auto' o 'disabled' para el campo share"
+  oc_002:
+    message: "La ruta de instruccion '%{path}' no existe y no es un patron glob valido"
+    suggestion: "Corrige o elimina la ruta de instruccion rota"
+  oc_003:
+    message: "Error de analisis en opencode.json: %{error}"
+    suggestion: "Corrige la sintaxis JSON en opencode.json"
+
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply locales/es.yml:374]: Fixed. Spanish locale updated with specific message keys matching the English ones.

File: locales/zh-CN.yml [hunk 3]
```diff
@@ -327,6 +327,17 @@ rules:
...
+  # --- OpenCode (opencode.rs) ---
+  oc_001:
+    message: "无效的共享模式 '%{value}'。有效值: 'manual', 'auto', 'disabled'"
+    suggestion: "使用 'manual'、'auto' 或 'disabled' 作为 share 字段的值"
+  oc_002:
+    message: "指令路径 '%{path}' 不存在且不是有效的 glob 模式"
+    suggestion: "修复或移除损坏的指令路径"
+  oc_003:
+    message: "opencode.json 解析错误: %{error}"
+    suggestion: "修复 opencode.json 中的 JSON 语法"
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply locales/zh-CN.yml]: Fixed. Chinese locale updated with specific message keys matching the English ones.

File: crates/agnix-core/src/schemas/opencode.rs [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+        .and_then(|v| v.as_str())
+        .map(|s| s.to_string());
+
+    let instructions = value.get("instructions").and_then(|v| {
+        v.as_array().map(|arr| {
+            arr.iter()
+                .filter_map(|item| item.as_str().map(|s| s.to_string()))
+                .collect()
+        })
+    });
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/schemas/opencode.rs:96]: Fixed. The schema parser now tracks `share_wrong_type` and `instructions_wrong_type` flags via the raw `serde_json::Value`. When `share` is present but not a string, OC-001 fires with a `type_error` message. When `instructions` is present but not an array (or contains non-string elements), OC-002 fires with a `type_error` message. Tests added for all scenarios.

File: crates/agnix-core/src/rules/opencode.rs [hunk 5]
```diff
@@ -0,0 +1,485 @@
...
+            if let Some(ref instructions) = schema.instructions {
+                let config_dir = path.parent().unwrap_or(Path::new("."));
+                let instructions_line = find_key_line(content, "instructions").unwrap_or(1);
+                let fs = config.fs();
+
+                for instruction_path in instructions {
+                    if instruction_path.trim().is_empty() {
+                        continue;
+                    }
+
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/rules/opencode.rs:104]: Fixed. When `instructions` is not an array or contains non-string items, OC-002 now fires with a `type_error` message. The `instructions_wrong_type` flag is set in the parser for both non-array types and arrays with non-string elements.

File: crates/agnix-core/src/rules/opencode.rs [hunk 6]
```diff
@@ -0,0 +1,485 @@
...
+                                path.to_path_buf(),
+                                instructions_line,
+                                0,
+                                "OC-002",
+                                t!("rules.oc_002.message", path = instruction_path.as_str()),
+                            )
+                            .with_suggestion(t!("rules.oc_002.suggestion")),
+                        );
+                        continue;
+                    }
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/rules/opencode.rs:121]: Fixed. OC-002 now uses four distinct i18n message keys: `not_found`, `invalid_glob`, `traversal`, and `type_error` - each accurately describing the specific failure mode.

File: crates/agnix-core/src/rules/opencode.rs [hunk 7]
```diff
@@ -0,0 +1,485 @@
...
+                            line,
+                            0,
+                            "OC-001",
+                            t!("rules.oc_001.message", value = share_value.as_str()),
+                        )
+                        .with_suggestion(t!("rules.oc_001.suggestion")),
+                    );
+                }
+            }
+        }
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/rules/opencode.rs:78]: Fixed. OC-001 now checks `share_wrong_type` flag before checking the string value. `{"share": true}` and `{"share": 123}` now trigger OC-001 with a `type_error` message. Tests added.

File: crates/agnix-core/src/rules/opencode.rs [hunk 8]
```diff
@@ -0,0 +1,420 @@
...
+                                .with_suggestion(t!("rules.oc_002.suggestion")),
+                            );
+                        }
+                        // Valid glob patterns are allowed even if no files match yet
+                        continue;
+                    }
+
+                    // For non-glob paths, check if the file exists relative to config dir
+                    let resolved = config_dir.join(instruction_path);
+                    if !resolved.exists() {
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/rules/opencode.rs]: Fixed. Path traversal is now explicitly rejected (absolute paths and '..' components) with a dedicated 'traversal' i18n message. The validator also uses config.fs() instead of direct filesystem calls.

File: crates/agnix-core/src/rules/opencode.rs [hunk 9]
```diff
@@ -0,0 +1,420 @@
...
+/// Find the 1-indexed line number of a JSON key in the content
+fn find_key_line(content: &str, key: &str) -> Option<usize> {
+    let needle = format!("\"{}\"", key);
+    for (i, line) in content.lines().enumerate() {
+        if line.contains(&needle) {
+            return Some(i + 1);
+        }
+    }
+    None
+}
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/rules/opencode.rs:182]: Fixed. find_key_line now checks that a colon follows the quoted key name (with optional whitespace), so it only matches JSON object keys, not string values containing the same text.

File: crates/agnix-core/src/schemas/opencode.rs [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+    pub line: usize,
+    pub column: usize,
+}
+
+/// Parse opencode.json content
+pub fn parse_opencode_json(content: &str) -> ParsedOpenCodeConfig {
+    // Try to strip JSONC comments before parsing
+    let stripped = strip_jsonc_comments(content);
+
+    match serde_json::from_str::<OpenCodeSchema>(&stripped) {
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/schemas/opencode.rs]: Fixed. The schema parser now uses a two-pass approach: first deserialize to serde_json::Value to detect type mismatches (share_wrong_type, instructions_wrong_type flags), then extract string/array values. Type mismatches now produce OC-001/OC-002 diagnostics instead of OC-003 parse errors.

File: crates/agnix-core/src/rules/opencode.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/rules/opencode.rs]: Fixed. The validator now uses config.fs() (line 98) and fs.exists() (line 146) instead of direct std::path::Path::exists(), which respects the FileSystem trait abstraction and enables MockFileSystem testing.

File: crates/agnix-core/src/schemas/opencode.rs [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+                while i + 1 < len && !(chars[i] == '*' && chars[i + 1] == '/') {
+                    // Preserve newlines for line counting
+                    if chars[i] == '\n' {
+                        result.push('\n');
+                    }
+                    i += 1;
+                }
+                if i + 1 < len {
+                    i += 2; // skip */
+                }
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/schemas/opencode.rs:157]: Acknowledged. The JSONC comment stripping replaces comment content with spaces to preserve line numbers, but column offsets within a line containing stripped comments can be slightly off. This is a known limitation shared with other validators that handle JSONC. The line number (which is the primary navigation target) remains accurate.

File: crates/agnix-core/src/rules/opencode.rs [hunk 12]
```diff
@@ -0,0 +1,485 @@
...
+    #[test]
+    fn test_type_mismatch_instructions_not_array() {
+        // "instructions": "README.md" is valid JSON but wrong type
+        let diagnostics = validate(r#"{"instructions": "README.md"}"#);
+        let oc_003: Vec<_> = diagnostics.iter().filter(|d| d.rule == "OC-003").collect();
+        assert!(
+            oc_003.is_empty(),
+            "Type mismatch should not be a parse error"
+        );
+    }
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/rules/opencode.rs:515]: Fixed. The type mismatch tests now assert that OC-001/OC-002 diagnostics are produced. See test_type_mismatch_share_not_string (line 467), test_type_mismatch_share_number (line 482), test_type_mismatch_instructions_not_array (line 489), and test_type_mismatch_instructions_with_non_string_elements (line 508).

File: crates/agnix-core/src/schemas/opencode.rs [hunk 13]
```diff
@@ -0,0 +1,293 @@
...
+    pub message: String,
+    pub line: usize,
+    pub column: usize,
+}
+
+/// Parse opencode.json content
+///
+/// Uses a two-pass approach: first validates JSON syntax with `serde_json::Value`,
+/// then extracts the typed schema. This ensures that type mismatches (e.g.,
+/// `"share": true`) are reported as OC-001/OC-002 issues rather than OC-003.
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/schemas/opencode.rs:51]: Fixed. The doc comment and implementation now match. The two-pass parsing approach correctly sets share_wrong_type and instructions_wrong_type flags, which the validator checks before string/array validation. Type mismatches produce OC-001/OC-002 diagnostics as documented.
